Fix OS error after changing the cluster when editing DetailsCard component

bamsalem opened the pull request:
This change should fix the issue that mentioned in this [comment](https://github.com/oVirt/ovirt-web-ui/issues/1320#issue-732507158).

The cause of the issue:
In `operatingSystems` there is no entry `'type'`:
![image](https://user-images.githubusercontent.com/64131213/102238182-2519df80-3efe-11eb-81b8-7740bd36414a.png)
As a result `templateOs` will always be `undefined`,
So the requested fix should be, to change it to `'name'`, but it won't work for ppc 64 architecture because any OS has its own id and name and it might cause errors,
for example Other OS for ppc64 has different id from x86_64.
So I used the `'description'` attribute (like I did in #1349 ).

Fixes: #1320

ovirt-infra commented:
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

rszwajko commented on src/components/VmDetails/cards/DetailsCard/index.js:
```diff
@@ -251,35 +257,37 @@ class DetailsCard extends React.Component {
             if (template) {
               const templateOsType = template.getIn(['os', 'type'], 'other')
-              const templateOs = this.props.operatingSystems.find(os => os.get('type') === templateOsType)
+              const oldOS = operatingSystems.find(os => os.get('name') === templateOsType)
```
IMHO this line contains the fix - `os.name` and `template.os.type` seem to be OS unique name.
Check both REST definitions:
1.  `os.name` is set [here](https://github.com/oVirt/ovirt-engine/blob/2314e3b31e42ff1c96395948af7a9a7f996f56ab/backend/manager/modules/restapi/jaxrs/src/main/java/org/ovirt/engine/api/restapi/resource/BackendOperatingSystemResource.java#L36)
2. `template.os.type` is set [here](https://github.com/oVirt/ovirt-engine/blob/2314e3b31e42ff1c96395948af7a9a7f996f56ab/backend/manager/modules/restapi/types/src/main/java/org/ovirt/engine/api/restapi/types/TemplateMapper.java#L90)

Why not just:
```
const templateOs = operatingSystems.find(os => os.get('name') === templateOsType) || findDefaultForCluster()
```
?
Of course `findDefaultForCluster()` remains the tricky part :)

rszwajko commented on src/components/VmDetails/cards/DetailsCard/index.js:
```diff
@@ -251,35 +257,37 @@ class DetailsCard extends React.Component {
-              const templateOs = this.props.operatingSystems.find(os => os.get('type') === templateOsType)
+              const oldOS = operatingSystems.find(os => os.get('name') === templateOsType)
+              const clustersOs = createOsList(updates.getIn(['cluster', 'id']), clusters, operatingSystems)
+              const templateOs = clustersOs.find(os => os.value === oldOS.get('description')) || clustersOs.find(os => os.value === 'Other OS')
```
As mentioned  [here](https://github.com/oVirt/ovirt-web-ui/pull/1349#discussion_r545376619) I think we should base the logic on unique names/IDs.

bamsalem replied:
I've sent you question on slack on this subject, I'll post the conclusions here later

sgratch commented:
@bamsalem @rszwajko 
Both OS name and OS id are unique and can be used. The default OS indexes are kept as constants on backend:
https://github.com/oVirt/ovirt-engine/blob/af7bb6b91639edf92b977b08484423cd0108d723/backend/manager/modules/common/src/main/java/org/ovirt/engine/core/common/osinfo/OsRepository.java#L22-L24
https://github.com/oVirt/ovirt-engine/blob/af7bb6b91639edf92b977b08484423cd0108d723/backend/manager/modules/utils/src/main/java/org/ovirt/engine/core/utils/OsRepositoryImpl.java#L52-L58

And the default OS per cluster is based on its architecture:
setOSType(new NotChangableForVmInPoolListModel<Integer>() {
 ...
            super.setSelectedItem(AsyncDataProvider.getInstance().getDefaultOs(cluster.getArchitecture()));
 ...

The main important thing here is to fix the error exception of course, but avoid handing a use case of switching between clusters from **different** architectures since this is not supported and will fail on backed anyway, so no need to handle that on web-ui as well.

sgratch commented:
ci add to whitelist

sgratch commented on src/components/VmDetails/cards/DetailsCard/index.js:
```diff
@@ -251,28 +266,32 @@ class DetailsCard extends React.Component {
-              const templateOs = this.props.operatingSystems.find(os => os.get('type') === templateOsType)
+              const clusterId = updates.getIn(['cluster', 'id'])
+              const clusterArchitecture = getClusterArchitecture(clusterId, clusters)
+              const templateOsName = template.getIn(['os', 'type'], 'other')
```
Why the const is **templateOsName** and not **templateOsType**?

bamsalem replied:
It's templateOsName because it's stored as `name` in the `operatingSystems` objects list:
![image](https://user-images.githubusercontent.com/64131213/102238182-2519df80-3efe-11eb-81b8-7740bd36414a.png)
Another reason, we compare it to OS Name in line [293](https://github.com/oVirt/ovirt-web-ui/blob/b63b0737f16f1354d2437560b9b55e7adb9eb92d/src/components/VmDetails/cards/DetailsCard/index.js#L293).
I'm not sure which one is clearer but if you think templateOsType is clearer I'll change that.

sgratch replied:
The right name should really be templateOsName ("type" term is only used for templates). For clarifications I suggest adding a line comment, specifying that template.os.type and operatingSystems.name are referring to the same value on backend.

sgratch reviewed (commented):
The original exception is not reproduced and therefore fixed, but saving the VM after changing to a cluster from the same architecture fails, even though it shouldn't, with an error:
"EDIT_VM failed [Cannot edit VM. The requested template doesn't belong to the same base template as the original template.]"

bamsalem commented on src/constants/oparatingSystems.js:
```diff
@@ -0,0 +1,5 @@
+export const defaultOperatingSystemIds = [
```
Done.

bamsalem commented:
> The original exception is not reproduced and therefore fixed, but saving the VM after changing to a cluster from the same architecture fails, even though it shouldn't, with an error:
> "EDIT_VM failed [Cannot edit VM. The requested template doesn't belong to the same base template as the original template.]"

I checked that and I got this error too, but I found also that cluster has been changed , after the error is shown click cancel and wait for the next refresh. you can see it's changed...
Do you want me to find out what is the cause and fix it? Or we should deal with that in another PR?

sgratch commented:
> I checked that and I got this error too, but I found also that cluster has been changed , after the error is shown click cancel and wait for the next refresh. you can see it's changed...

Yes, cluster is changed and template is switched to blank but this error shouldn't occur. 
The reason for this error is that if cluster is changed by the user for vm based on non Blank template then auto changing the template is not allowed since this vm was provisioned from that template (template disks are not cloned) so you can't just change the template without cloning the disks.
The right behaviour is therefore to allow changing cluster (if allowed by architecture) and doesn't auto change template and OS type to Blank, "other os". Only the cluster should be changed. This is how it behaves on webadmin as well for vms based on non Blank template.

> Do you want me to find out what is the cause and fix it? Or we should deal with that in another PR?

You can but you'll anyway need to fix this PR as well.

bamsalem commented:
@sgratch 
I removed the template and os change on cluster change, I also added option to filter the clusterList by architecture to show only clusters with the same architecture (clusters that the vm can migrate to)

sgratch reviewed (approved):
LGTM

isaranova commented:
LGTM
